smithy-model/src/main/java/software/amazon/smithy/model/knowledge/ServiceIndex.java

```diff
@@ -198,6 +200,41 @@ public Map<ShapeId, Trait> getEffectiveAuthSchemes(ToShapeId service, ToShapeId
                 .orElse(Collections.emptyMap());
     }
 
+    /**
+     * Gets a list of effective authentication schemes applied to an operation bound within a service, that also
+     * includes the {@code smithy.api#noAuth} scheme when no authentication is to be used.
+     *
+     * <p>This behaves the same as {@link #getEffectiveAuthSchemes(ToShapeId, ToShapeId)} with the following
+     * differences:
+     * If the operation has no effective auth schemes, instead of an empty map, it returns the {@code smithy.api#noAuth}
+     * auth scheme.
+     * If the operation has the {@code smithy.api#optionalAuth} trait, it adds {@code smithy.api#noAuth} to the end.
+     *
+     * <p>This is a convenience wrapper around getEffectiveAuthSchemes, that accounts for
+     * {@code smithy.api#optionalAuth} and avoids having to special case handling empty result. The returned map will
+     * always contain at least 1 entry.
+     *
+     * <p>The {@code smithy.api#noAuth} scheme, if present, is always the last scheme.
+     *
+     * @param service Service the operation is within.
+     * @param operation Operation to get the effective authentication schemes of.
+     * @return Returns a map of the trait shape ID to the auth trait itself.
+     */
+    public Map<ShapeId, Trait> getNoAuthAwareEffectiveAuthSchemes(ToShapeId service, ToShapeId operation) {
```

Comment:
I was wondering if overloading `getEffectiveAuthSchemes(ToShapeId, ToShapeId)` with a mode that would denote adding `@noAuth` could work. E.g. `getEffectiveAuthSchemes(ToShapeId, ToShapeId, AuthMode)`, where `AuthMode` could either be `DEFAULT` (which would be existing behavior) and `NO_AUTH_AWARE`.

This way, we can still use the reasonable existing method name `getEffectiveAuthSchemes` and be backwards-compatible with existing users.

Also, this would need an equivalent `getNoAuthAwareEffectiveAuthSchemes(ToShapeId service)` that only gets the auth schemes on a service shape.